model-engine/model_engine_server/infra/gateways/gcs_llm_artifact_gateway.py

```diff
@@ -0,0 +1,115 @@
+import json
+import os
+from typing import Any, Dict, List
+
+from google.cloud import storage
+from model_engine_server.common.config import get_model_cache_directory_name, hmi_config
+from model_engine_server.core.loggers import logger_name, make_logger
+from model_engine_server.core.utils.url import parse_attachment_url
+from model_engine_server.domain.gateways import LLMArtifactGateway
+
+logger = make_logger(logger_name())
+
+
+class GCSLLMArtifactGateway(LLMArtifactGateway):
+    """
+    Concrete implementation for interacting with a filesystem backed by GCS.
+    """
+
+    def _get_gcs_client(self, kwargs) -> storage.Client:
+        """
+        Returns a GCS client. If desired, you could pass in project info
+        or credentials via `kwargs`.
+        """
+        project = kwargs.get("gcp_project", os.getenv("GCP_PROJECT"))
+        return storage.Client(project=project)
+
+    def list_files(self, path: str, **kwargs) -> List[str]:
+        """
+        Lists all files under the path argument in GCS. The path is expected
+        to be in the form 'gs://bucket/prefix'.
+        """
+        gcs = self._get_gcs_client(kwargs)
+        parsed_remote = parse_attachment_url(path, clean_key=False)
+        bucket_name = parsed_remote.bucket
+        prefix = parsed_remote.key
+
+        bucket = gcs.bucket(bucket_name)
+        files = [blob.name for blob in bucket.list_blobs(prefix=prefix)]
+        return files
+
+    def download_files(self, path: str, target_path: str, overwrite=False, **kwargs) -> List[str]:
+        """
+        Downloads all files under the given path to the local target_path directory.
+        """
+        gcs = self._get_gcs_client(kwargs)
+        parsed_remote = parse_attachment_url(path, clean_key=False)
+        bucket_name = parsed_remote.bucket
+        prefix = parsed_remote.key
+
+        bucket = gcs.bucket(bucket_name)
+        blobs = bucket.list_blobs(prefix=prefix)
+        downloaded_files = []
+
+        for blob in blobs:
+            # Remove prefix and leading slash to derive local name
+            file_path_suffix = blob.name.replace(prefix, "").lstrip("/")
```

Comment:
do you want to `replace(prefix, "", count=1)`? (or something like that) just in case the prefix appears elsewhere in the string

for that matter is that also a bug in the s3 implementation?